src/librustc_borrowck/borrowck/mod.rs

```diff
@@ -62,27 +66,51 @@ pub struct LoanDataFlowOperator;
 
 pub type LoanDataFlow<'a, 'tcx> = DataFlowContext<'a, 'tcx, LoanDataFlowOperator>;
 
-pub fn check_crate<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>) {
-    for body_owner_def_id in tcx.body_owners() {
-        tcx.borrowck(body_owner_def_id);
+pub fn check_crate<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>, save_analysis: bool) -> Option<HashMap<DefId, AnalysisResult<'a, 'tcx>>> {
+    // FIXME: nashenas88 support this through TyCtxt
+    if save_analysis {
+        let mut map = HashMap::new();
+        for body_owner_def_id in tcx.body_owners() {
+            let result = borrowck(tcx, body_owner_def_id, save_analysis).unwrap();
+            map.insert(body_owner_def_id, result);
+        }
+        Some(map)
+    } else {
+        for body_owner_def_id in tcx.body_owners() {
+            tcx.borrowck(body_owner_def_id);
+        }
+        None
     }
 }
 
 pub fn provide(providers: &mut Providers) {
     *providers = Providers {
-        borrowck,
+        borrowck: borrowck_provider,
         ..*providers
     };
 }
 
 /// Collection of conclusions determined via borrow checker analyses.
 pub struct AnalysisData<'a, 'tcx: 'a> {
+    pub safe_loans: Vec<SafeLoan>,
     pub all_loans: Vec<Loan<'tcx>>,
     pub loans: DataFlowContext<'a, 'tcx, LoanDataFlowOperator>,
     pub move_data: move_data::FlowedMoveData<'a, 'tcx>,
 }
 
-fn borrowck<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>, owner_def_id: DefId) {
+pub struct AnalysisResult<'a, 'tcx: 'a> {
+    pub bccx: BorrowckCtxt<'a, 'tcx>,
```

Comment:
I remember bringing this up in IRC at some point (not `BorrowckCtxt` specifically, but using internals of the borrow checker), but I can't recall exactly who I was speaking with. The issue was that this is used to calculate some spans for assignments, but I don't know which ones to compute it for until later, and the worry was that computing it early could just be unnecessary computation for when someone isn't using the save context.

This might change entirely with the NLL approach though. For reference, it's used in `dump_visitor.rs`, `fn process_assignment` line 222.